### PR TITLE
squshed commits from leca/amd

### PR DESCRIPTION
**Pull request by jslhcl**
### Description
support amd gpu ep on multi-stream

### Motivation and Context
support amd gpu ep on multi-stream

### CODE REVIEW

File: onnxruntime/core/providers/migraphx/gpu_data_transfer.h [hunk 1]
```diff
@@ -8,30 +8,15 @@
...
-  kHipStreamCopyIn,
-  kHipStreamCopyOut,
-  kTotalHipStreams,
-};
 
 class GPUDataTransfer : public IDataTransfer {
  public:
-  GPUDataTransfer(hipStream_t stream);
-  ~GPUDataTransfer();
+  GPUDataTransfer() {};
```
Comment (RandySheriffH) [inline onnxruntime/core/providers/migraphx/gpu_data_transfer.h:14]: >{}; [](http://example.com/codeflow?start=20&length=3)

= default;

File: onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc [hunk 2]
```diff
@@ -115,6 +119,26 @@ MIGraphXExecutionProvider::MIGraphXExecutionProvider(const MIGraphXExecutionProv
...
+  ROCBLAS_CALL_THROW(rocblas_create_handle(&external_rocblas_handle_));
+  ROCBLAS_CALL_THROW(rocblas_set_stream(external_rocblas_handle_, stream_));
+
+  MIOPEN_CALL_THROW(miopenCreate(&external_miopen_handle_));
+  MIOPEN_CALL_THROW(miopenSetStream(external_miopen_handle_, stream_));
+}
+
+MIGraphXExecutionProvider::~MIGraphXExecutionProvider() {
+  try {
+    ROCBLAS_CALL_THROW(rocblas_destroy_handle(external_rocblas_handle_));
```
Comment (RandySheriffH) [inline onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc:132]: >ROCBLAS_CALL_THROW [](http://example.com/codeflow?start=4&length=18)

Is it guaranteed ROCBLAS_CALL_THROW will give out a std exception so we could capture? 
Further, maybe in destructor we should simply log the call error of "rocblas_destroy_handle" instead using ROCBLAS_CALL_THROW.

File: onnxruntime/core/providers/rocm/gpu_data_transfer.h [hunk 3]
```diff
@@ -8,32 +8,17 @@
...
-  kHipStreamCopyIn,
-  kHipStreamCopyOut,
-  kTotalHipStreams,
-};
-
 class GPUDataTransfer : public IDataTransfer {
  public:
-  GPUDataTransfer(hipStream_t stream, bool do_copy_in_default_stream = true);
-  ~GPUDataTransfer();
+  GPUDataTransfer() {};
```
Comment (RandySheriffH) [inline onnxruntime/core/providers/rocm/gpu_data_transfer.h:13]: >{}; [](http://example.com/codeflow?start=20&length=3)

= default;

File: onnxruntime/core/providers/rocm/rocm_stream_handle.cc [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    HIP_CALL_THROW(hipEventCreateWithFlags(&event_, hipEventDisableTiming));
+  }
+
+  ~RocmNotification() {
+    if (event_)
+      HIP_CALL_THROW(hipEventDestroy(event_));
+  }
+
+  void Activate() override {
+    // record event with hipEventBlockingSync so we can support sync on host with out busy wait.
```
Comment (RandySheriffH) [inline onnxruntime/core/providers/rocm/rocm_stream_handle.cc:18]: >ith out  [](http://example.com/codeflow?start=78&length=8)

nit: grammar

File: onnxruntime/core/providers/rocm/rocm_stream_handle.cc [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    HIP_CALL_THROW(hipEventRecord(event_, static_cast<hipStream_t>(stream_.GetHandle())));
+  }
+
+  void wait_on_device(Stream& device_stream) {
+    ORT_ENFORCE(device_stream.GetDevice().Type() == OrtDevice::GPU);
```
Comment (RandySheriffH) [inline onnxruntime/core/providers/rocm/rocm_stream_handle.cc:23]: >ORT_ENFORCE(device_stream.GetDevice().Type() == OrtDevice::GPU); [](http://example.com/codeflow?start=4&length=64)

Always report a text message along with enforce.

File: onnxruntime/core/providers/rocm/rocm_stream_handle.cc [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+  delete info;
+}
+
+Status RocmStream::CleanUpOnRunEnd() {
+  if (deferred_cpu_buffers_.empty())
+    return Status::OK();
+  // Release the ownership of cpu_buffers_info so that the underlying
+  // object will keep alive until the end of ReleaseCpuBufferCallback.
+  if (release_cpu_buffer_on_rocm_stream_ && cpu_allocator_->Info().alloc_type == OrtArenaAllocator) {
+    auto cpu_buffers_info = new CpuBuffersInfo;
```
Comment (RandySheriffH) [inline onnxruntime/core/providers/rocm/rocm_stream_handle.cc:106]: >auto cpu_buffers_info = new CpuBuffersInfo; [](http://example.com/codeflow?start=4&length=43)

"new" here violate ort compliance, pls always be with smart pointers.

File: onnxruntime/core/providers/rocm/rocm_stream_handle.cc [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    cpu_buffers_info->allocator = cpu_allocator_;
+    cpu_buffers_info->buffers = new void*[deferred_cpu_buffers_.size()];
+    for (size_t i = 0; i < deferred_cpu_buffers_.size(); ++i) {
+      cpu_buffers_info->buffers[i] = deferred_cpu_buffers_.at(i);
+    }
+    cpu_buffers_info->n_buffers = deferred_cpu_buffers_.size();
```
Comment (RandySheriffH) [inline onnxruntime/core/providers/rocm/rocm_stream_handle.cc:112]: Put the logic in the constructor, this is protocol must be respected between caller and the struct.

File: onnxruntime/core/providers/rocm/rocm_stream_handle.h [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+  bool own_stream_{true};
+
+  miopenHandle_t miopen_handle_{};
+
+  rocblas_handle rocblas_handle_{};
+
+ private:
+  std::vector<void*> deferred_cpu_buffers_;
+  AllocatorPtr cpu_allocator_;
```
Comment (RandySheriffH) [inline onnxruntime/core/providers/rocm/rocm_stream_handle.h:37]: should we allow move or are the default move expected to these members?

File: onnxruntime/core/providers/rocm/rocm_stream_handle.h [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+             const OrtDevice& device,
+             AllocatorPtr cpu_allocator,
+             bool release_cpu_buffer_on_rocm_stream,
+             bool own_flag,
+             miopenHandle_t external_miopen_handle,
+             rocblas_handle external_rocblas_handle);
+
+  ~RocmStream();
+
+  std::unique_ptr<synchronize::Notification> CreateNotification(size_t /*num_consumers*/) override;
```
Comment (RandySheriffH) [inline onnxruntime/core/providers/rocm/rocm_stream_handle.h:21]: >CreateNotification(size_t /*num_consumers*/) override; [](http://example.com/codeflow?start=45&length=54)

nit: pls align the style - for function declaration, letting there be a var name will not trigger any warning.

File: onnxruntime/core/providers/rocm/rocm_execution_provider.cc [hunk 10]
```diff
@@ -312,134 +310,30 @@ void ROCMExecutionProvider::ReleasePerThreadContext() const {
...
-  // Pinned memory allocator is shared between threads, but ROCM memory allocator is per-thread or it may cause result changes
-  // A hypothesis is that arena allocator is not aligned with ROCM output cache, and data from different kernel writes may
-  // cause cacheline to contain dirty data.
   if (mem_type == OrtMemTypeDefault) {
-    return GetPerThreadContext().GetAllocator();
-  } else {
-    return IExecutionProvider::GetAllocator(id, mem_type);
+    auto rocm_alloc = IExecutionProvider::GetAllocator(id, mem_type);
+    if (!rocm_alloc) {
+      return CreateRocmAllocator(info_.device_id, info_.gpu_mem_limit, info_.arena_extend_strategy,
```
[the remark this answers is not in the document]
Comment (jslhcl) [reply onnxruntime/core/providers/rocm/rocm_execution_provider.cc:316]: handled in CreateRocmAllocator()

File: onnxruntime/core/providers/rocm/rocm_execution_provider.cc [hunk 11]
```diff
@@ -114,27 +114,27 @@ AllocatorPtr ROCMExecutionProvider::CreateRocmAllocator(OrtDevice::DeviceId devi
...
 
     // ROCM malloc/free is expensive so always use an arena
     return CreateAllocator(default_memory_info);
   }
 }
 
-ROCMExecutionProvider::PerThreadContext::PerThreadContext(OrtDevice::DeviceId device_id, hipStream_t stream, size_t gpu_mem_limit,
-                                                          ArenaExtendStrategy arena_extend_strategy, ROCMExecutionProviderExternalAllocatorInfo external_allocator_info,
-                                                          OrtArenaCfg* default_memory_arena_cfg) {
+ROCMExecutionProvider::PerThreadContext::PerThreadContext(OrtDevice::DeviceId device_id, hipStream_t stream, size_t /*gpu_mem_limit*/,
```
[the remark this answers is not in the document]
Comment (jslhcl) [reply onnxruntime/core/providers/rocm/rocm_execution_provider.cc:128]: Yes, it is used in line 134 and line 137

File: onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc (hunk 2 above)
Comment (jslhcl) [reply onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc:132]: it throws "ORT_THROW" just as CudaCall() does. 

Will follow the destructor of TensorRTExecutionProvider.

File: onnxruntime/contrib_ops/rocm/bert/attention.cc [hunk 12]
```diff
@@ -77,14 +77,14 @@ Status Attention<T>::ComputeInternal(OpKernelContext* context) const {
...
-  rocblas_handle rocblas = RocblasHandle();
+  rocblas_handle rocblas = GetRocblasHandle(context);
   constexpr size_t element_size = sizeof(T);
 
   // Use GEMM for fully connection.
   int m = batch_size * sequence_length;
   int n = 3 * hidden_size;
   int k = input_hidden_size;
-  auto gemm_buffer = GetScratchBuffer<T>(batch_size * sequence_length * 3 * hidden_size * element_size);
+  auto gemm_buffer = GetScratchBuffer<T>(batch_size * sequence_length * 3 * hidden_size * element_size, context->GetComputeStream());
```
[the remark this answers is not in the document]
Comment (jslhcl) [reply onnxruntime/contrib_ops/rocm/bert/attention.cc:87]: You mean replace OrtStream(context) with context->GetComputeStream()? I think we should. I will do the change

File: onnxruntime/core/providers/rocm/rocm_stream_handle.cc (hunk 6 above)
Comment (jslhcl) [reply onnxruntime/core/providers/rocm/rocm_stream_handle.cc:106]: Will do the change on cuda part first

File: onnxruntime/core/providers/rocm/rocm_stream_handle.h (hunk 8 above)
Comment (jslhcl) [reply onnxruntime/core/providers/rocm/rocm_stream_handle.h:37]: disallow move contructor

File: onnxruntime/core/providers/rocm/rocm_stream_handle.h (hunk 9 above)
Comment (jslhcl) [reply onnxruntime/core/providers/rocm/rocm_stream_handle.h:21]: keep num_consumers

File: onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc (hunk 2 above)
Comment (jslhcl) [reply onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc:132]: Do not call ROCBLAS_CALL_THROW

File: onnxruntime/core/providers/rocm/rocm_stream_handle.cc (hunk 6 above)
Comment (souptc) [reply onnxruntime/core/providers/rocm/rocm_stream_handle.cc:107]: The code is moved from here in main branch:
https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/cuda/cuda_execution_provider.cc#L381

The implementation is updated in main branch, we need to re-sync, in both cuda and AMD ep.

**Pull request merged**